### PR TITLE
Raspberry Pi Pico Clock Control Driver

### PR DESCRIPTION
**Pull request by andrei-edward-popa**
I have added driver support for clock control for Raspberry Pi Pico. Currently external clock sources are not supported by this driver.

### CODE REVIEW

File: dts/bindings/clock/raspberrypi,pico-syscon.yaml [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# Copyright (c) 2022 Andrei-Edward Popa
+# SPDX-License-Identifier: Apache-2.0
+
+description: Raspberry Pi Pico clock controller node
+
+compatible: "raspberrypi,pico-syscon"
```
Comment (soburi) [inline dts/bindings/clock/raspberrypi,pico-syscon.yaml:6]: Is syscon the preferred postfix?
`syscon.yaml` exists, but it is not using that definition.
IMO, `clock-controller` (or `clock` related) postfix seems better here.

**Comment by MaureenHelm** [comment]
@andrei-edward-popa @yonsch ping

this also needs a rebase

File: dts/bindings/clock/raspberrypi,pico-syscon.yaml
Comment (yonsch) [inline dts/bindings/clock/raspberrypi,pico-syscon.yaml:1]: Why is the binding name `syscon`? It would be better to just call it clock/clk/clocks, like it's called in the datasheet

File: drivers/clock_control/clock_control_rpi_pico.c [hunk 2]
```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright (c) 2022 Andrei-Edward Popa
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT raspberrypi_pico_syscon
+
+#include <zephyr/drivers/clock_control/rpi_pico_clock_control.h>
```
Comment (yonsch) [inline drivers/clock_control/clock_control_rpi_pico.c:9]: This is the only place this header is used (as far as I can see), can it be removed and its contents be copied here?

File: dts/arm/rpi_pico/rp2040.dtsi [hunk 3]
```diff
@@ -68,6 +57,34 @@
...
+			reg = <0x40008000 DT_SIZE_K(4)>;
+			#clock-cells = <1>;
+			status = "okay";
+			ref-frequency = <12000000>;
+			sys-frequency = <125000000>;
+			peri-frequency = <125000000>;
+			usb-frequency = <48000000>;
+			adc-frequency = <48000000>;
+			rtc-frequency = <46875>;
+			xtal_clk: xtal-clk {
```
Comment (yonsch) [inline dts/arm/rpi_pico/rp2040.dtsi:71]: Are these still needed?

File: drivers/watchdog/wdt_rpi_pico.c [hunk 4]
```diff
@@ -154,18 +154,19 @@ static const struct wdt_driver_api wdt_rpi_pico_driver_api = {
...
-		.load = (CONFIG_WDT_RPI_PICO_INITIAL_TIMEOUT *                                     \
-			 RPI_PICO_WDT_TIME_MULTIPLICATION_FACTOR),                                 \
-		.enabled = false                                                                   \
-	};                                                                                         \
-	DEVICE_DT_DEFINE(DT_NODELABEL(wdt##idx), wdt_rpi_pico_init, NULL, &wdt_##idx##_data,       \
-			 &wdt_##idx##_config, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,    \
+#define WDT_RPI_PICO_WDT_DEVICE(idx)                                                             \
+	static const struct wdt_rpi_pico_config wdt_##idx##_config = {                           \
+		.xtal_frequency =                                                                \
+			DT_PROP(DT_CHILD(DT_NODELABEL(syscon), xtal_clk), clock_frequency)       \
```
Comment (yonsch) [inline drivers/watchdog/wdt_rpi_pico.c:160]: Shouldn't that be `clk_ref`?

File: drivers/serial/uart_rpi_pico.c [hunk 5]
```diff
@@ -59,13 +60,51 @@ static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
...
+	}
+
+	ret = clock_control_get_rate(cfg->clk_dev, (clock_control_subsys_t)cfg->clk_id, &pclk);
+	if (ret < 0 || pclk == 0) {
+		return -EINVAL;
+	}
+
+	baudrate_div = (8 * pclk / input_baudrate);
+	baudrate_int = baudrate_div >> 7;
+	baudrate_frac = baudrate_int == 0 || baudrate_int >= 65535 ? 0 :
```
Comment (yonsch) [inline drivers/serial/uart_rpi_pico.c:85]: Add parentheses for clarity (also for the next statement)

File: drivers/serial/uart_rpi_pico.c [hunk 6]
```diff
@@ -59,13 +60,51 @@ static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
...
+		((baudrate_div & 0x7f) + 1) / 2;
+	baudrate_int = baudrate_int == 0 ? 1 :
+		baudrate_int >= 65535 ? 65535 : baudrate_int;
+
+	uart_hw->ibrd = baudrate_int;
+	uart_hw->fbrd = baudrate_frac;
+
+	uart_hw->lcr_h |= 0;
```
Comment (yonsch) [inline drivers/serial/uart_rpi_pico.c:93]: Why is this needed?

File: drivers/serial/uart_rpi_pico.c
Comment (yonsch) [inline drivers/serial/uart_rpi_pico.c:1]: Now that we don't use picosdk for the serial driver, you can remove [this line](https://github.com/zephyrproject-rtos/zephyr/blob/44a8b70cccafe312a1a091de48a5ba204dadd6bb/drivers/serial/Kconfig.rpi_pico#L9)

**Review by yonsch — CHANGES_REQUESTED** [review]
Sorry for the delay, added a few comments. @soburi @nordic-krch can you take a look?

File: dts/bindings/clock/raspberrypi,pico-syscon.yaml (hunk 1 above)
Comment (andrei-edward-popa) [reply dts/bindings/clock/raspberrypi,pico-syscon.yaml]: Yeah, I think you are right, I used clocks now

File: dts/bindings/clock/raspberrypi,pico-syscon.yaml
Comment (andrei-edward-popa) [reply dts/bindings/clock/raspberrypi,pico-syscon.yaml:1]: Yes, you are right, I used clocks now

File: drivers/clock_control/clock_control_rpi_pico.c (hunk 2 above)
Comment (andrei-edward-popa) [reply drivers/clock_control/clock_control_rpi_pico.c:9]: A user can use the structure rpi_pico_clk_setup to setup a clock inside its application

File: dts/arm/rpi_pico/rp2040.dtsi (hunk 3 above)
Comment (andrei-edward-popa) [reply dts/arm/rpi_pico/rp2040.dtsi:75]: Yes, this are used inside the driver code

File: drivers/serial/uart_rpi_pico.c (hunk 5 above)
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c]: Done

File: drivers/watchdog/wdt_rpi_pico.c (hunk 4 above)
Comment (andrei-edward-popa) [reply drivers/watchdog/wdt_rpi_pico.c]: Yes, I read the docs and it seems to be clk_ref

File: drivers/serial/uart_rpi_pico.c
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c:1]: We cannot do this because some macros are from the Pico SDK

File: drivers/serial/uart_rpi_pico.c (hunk 6 above)
Comment (andrei-edward-popa) [reply drivers/serial/uart_rpi_pico.c:93]: PL011 needs a dummy line control register write to latch in the divisors

File: dts/arm/rpi_pico/rp2040.dtsi (hunk 3 above)
Comment (yonsch) [reply dts/arm/rpi_pico/rp2040.dtsi:75]: Okay, I see you're using them as `source_rate`. But in any case, these should be defined at the board level, as they depend on the value of the oscillator on the board (if it has one). What about choosing external clock sources and setting up the PLLs?

File: drivers/serial/uart_rpi_pico.c (hunk 6 above)
Comment (yonsch) [reply drivers/serial/uart_rpi_pico.c:93]: So please add a comment explaining it

File: drivers/clock_control/CMakeLists.txt [hunk 7]
```diff
@@ -21,7 +21,7 @@ zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_NRF                 clock_cont
 zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_NRF_DRIVER_CALIBRATION nrf_clock_calibration.c)
 zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_RV32M1_PCC          clock_control_rv32m1_pcc.c)
 zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_INFINEON_CAT1       clock_control_ifx_cat1.c)
-
```
Comment (yonsch) [inline drivers/clock_control/CMakeLists.txt:24]: Something happened here with the blank lines...

File: drivers/serial/uart_rpi_pico.c [hunk 8]
```diff
@@ -59,13 +60,51 @@ static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
...
+	}
+
+	ret = clock_control_get_rate(cfg->clk_dev, (clock_control_subsys_t)cfg->clk_id, &pclk);
+	if (ret < 0 || pclk == 0) {
+		return -EINVAL;
+	}
+
+	baudrate_div = (8 * pclk / input_baudrate);
+	baudrate_int = baudrate_div >> 7;
+	baudrate_frac = (baudrate_int == 0) || (baudrate_int >= 65535) ? 0 :
```
Comment (yonsch) [inline drivers/serial/uart_rpi_pico.c:85]: Add even more parentheses

File: drivers/clock_control/clock_control_rpi_pico.c [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+		[RPI_PICO_CLOCK_SYS] = {
+			.source = CLOCKS_CLK_SYS_CTRL_SRC_VALUE_CLKSRC_CLK_SYS_AUX,
+			.aux_source = CLOCKS_CLK_SYS_CTRL_AUXSRC_VALUE_CLKSRC_PLL_SYS,
+			.source_rate =  FIXED_CLOCK_FREQ(DT_NODELABEL(clocks), sys_pll),
+			.rate = DT_INST_PROP(0, sys_frequency)
+		},
+		[RPI_PICO_CLOCK_PERI] = {
+			.source = 0,
+			.aux_source = CLOCKS_CLK_PERI_CTRL_AUXSRC_VALUE_CLK_SYS,
+			.source_rate = DT_INST_PROP(0, sys_frequency),
```
Comment (yonsch) [inline drivers/clock_control/clock_control_rpi_pico.c:230]: Is this correct?

File: drivers/clock_control/clock_control_rpi_pico.c [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+
+#include <zephyr/drivers/clock_control/rpi_pico_clock_control.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/reset.h>
+
+/* pico-sdk includes */
+#include <hardware/clocks.h>
+
+#define CTRL_ENABLE_BITS CLOCKS_CLK_GPOUT0_CTRL_ENABLE_BITS
+#define FIXED_CLOCK_FREQ(node_id, child) DT_PROP(DT_CHILD(node_id, child), clock_frequency)
```
Comment (yonsch) [inline drivers/clock_control/clock_control_rpi_pico.c:17]: This seems to be written specifically to access the children of `clocks`, so maybe it should be rewritten to something like:
```
#define FIXED_CLOCK_FREQ(clock) DT_PROP(DT_CHILD(DT_NODELABEL(clocks), clock), clock_frequency)
```

**Review by yonsch — CHANGES_REQUESTED** [review]
A few questions, most importantly: how will clock sources be supported?
